### PR TITLE
fix: handle consecutive backslashes in `no-reference-like-urls`

### PR DESCRIPTION
**Pull request by lumirlumir**
## Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/.github/blob/master/CONTRIBUTING.md).

### Which language are you using?

CommonMark and GFM.

### What did you do?

I've tried using consecutive backslashes in the `no-reference-like-urls` rule, but it reported false positives as shown below:

#### First Problem

The code below doesn't report, which is the correct behavior.

```md
\[Mercury](mercury)

[mercury]: https://example.com/mercury
```

However, the code below does report, which are false positives:

```md
\\\[Mercury](mercury) <!-- False Positive -->

\\\\\[Mercury](mercury) <!-- False Positive -->

\\\\\\\[Mercury](mercury) <!-- False Positive -->

[mercury]: https://example.com/mercury
```

#### Second Problem

The code below is reported as a `link` type, which is the correct behavior.

```md
\![Mercury](mercury)

[mercury]: https://example.com/mercury
```

However, The code below is reported as an `image` type, which are incorrect behaviors.

The code below should be reported as a `link` type, since the `!` is escaped with a backslash.

```md
\\\![Mercury](mercury) <!-- Incorrect -->

\\\\\![Mercury](mercury) <!-- Incorrect -->

\\\\\\\![Mercury](mercury) <!-- Incorrect -->

[mercury]: https://example.com/mercury
```

### Link to minimal reproducible Example

Adding the following test cases would help identify the issues:

(Test cases not mentioned here but added were included just to verify the behavior.)

- `valid`:

```js
		`${"\\".repeat(3)}[Mercury](mercury)\n\n[mercury]: https://example.com/mercury`,
		`${"\\".repeat(5)}[Mercury](mercury)\n\n[mercury]: https://example.com/mercury`,
		`${"\\".repeat(7)}[Mercury](mercury)\n\n[mercury]: https://example.com/mercury`,
```

- `invalid`

```js
		{
			code: `${"\\".repeat(3)}![Mercury](mercury)\n\n[mercury]: https://example.com/mercury`,
			output: `${"\\".repeat(3)}![Mercury][mercury]\n\n[mercury]: https://example.com/mercury`,
			errors: [
				{
					messageId: "referenceLikeUrl",
					data: { type: "link", prefix: "" },
					line: 1,
					column: 5,
					endLine: 1,
					endColumn: 23,
				},
			],
		},
		{
			code: `${"\\".repeat(5)}![Mercury](mercury)\n\n[mercury]: https://example.com/mercury`,
			output: `${"\\".repeat(5)}![Mercury][mercury]\n\n[mercury]: https://example.com/mercury`,
			errors: [
				{
					messageId: "referenceLikeUrl",
					data: { type: "link", prefix: "" },
					line: 1,
					column: 7,
					endLine: 1,
					endColumn: 25,
				},
			],
		},
		{
			code: `${"\\".repeat(7)}![Mercury](mercury)\n\n[mercury]: https://example.com/mercury`,
			output: `${"\\".repeat(7)}![Mercury][mercury]\n\n[mercury]: https://example.com/mercury`,
			errors: [
				{
					messageId: "referenceLikeUrl",
					data: { type: "link", prefix: "" },
					line: 1,
					column: 9,
					endLine: 1,
					endColumn: 27,
				},
			],
		},
```		

## What changes did you make? (Give an overview)

This PR is a follow-up to #490.

In this PR, I've addressed false positives in `no-reference-like-urls` rule that were caused by consecutive backslash handling.

Previously, we used the following regex pattern to avoid matching backslash escapes:

- `(?<!(?<!\\)\\)` 

However, this pattern is incorrect because they don't properly handle consecutive backslash escapes metioned above.

So, I've updated them to use the following regex correctly:

- `(?<=(?<!\\)(?:\\{2})*)`

```js
(?<=(?<!\\)(?:\\{2})*)a
```

```txt
a           // even
\a          // odd
\\a         // even
\\\a        // odd
\\\\a       // even
\\\\\a      // odd
\\\\\\a     // even
```

image

image

## Related Issues

Ref: https://github.com/eslint/markdown/pull/490

## Is there anything you'd like reviewers to focus on?

In JavaScript, `\\\\\\\\` (8 backslashes) is interpreted as `\\\\` (4 backslashes) in Markdown, and is ultimately recognized as `\\` (2 backslashes). In this case, an even number of backslashes is not treated as an escape character and is recognized as a normal backslash.

### CODE REVIEW

File: src/rules/no-reference-like-urls.js [hunk 1]
```diff
@@ -16,7 +16,7 @@ import { findOffsets } from "../util.js";
 
 /**
  * @import { SourceRange } from "@eslint/core"
- * @import { Heading, Node, Paragraph, TableCell } from "mdast";
+ * @import { Heading, Paragraph, TableCell } from "mdast";
```
Comment (lumirlumir) [inline src/rules/no-reference-like-urls.js:19]: The `Node` type was not being used, so I removed it.

File: src/rules/no-reference-like-urls.js [hunk 2]
```diff
@@ -29,7 +29,7 @@ import { findOffsets } from "../util.js";
 
 /** Pattern to match both inline links: `[text](url)` and images: `![alt](url)`, with optional title */
 const linkOrImagePattern =
-	/(?<!(?<!\\)\\)(?<imageBang>!)?\[(?<label>(?:\\.|[^()\\]|\([\s\S]*\))*?)\]\((?<destination>[ \t]*(?:\r\n?|\n)?(?<![ \t])[ \t]*(?:<[^>]*>|[^ \t()]+))(?:[ \t]*(?:\r\n?|\n)?(?<![ \t])[ \t]*(?<title>"[^"]*"|'[^']*'|\([^)]*\)))?[ \t]*(?:\r\n?|\n)?(?<![ \t])[ \t]*\)(?!\()/gu;
+	/(?<=(?<!\\)(?:\\{2})*)(?<imageBang>!)?\[(?<label>(?:\\.|[^()\\]|\([\s\S]*\))*?)\]\((?<destination>[ \t]*(?:\r\n?|\n)?(?<![ \t])[ \t]*(?:<[^>]*>|[^ \t()]+))(?:[ \t]*(?:\r\n?|\n)?(?<![ \t])[ \t]*(?<title>"[^"]*"|'[^']*'|\([^)]*\)))?[ \t]*(?:\r\n?|\n)?(?<![ \t])[ \t]*\)(?!\()/gu;
```
Comment (mdjermanovic) [inline src/rules/no-reference-like-urls.js:32]: Is there a reason this rule parses paragraphs and other nodes to search for links and images instead of checking `link` and `image` nodes directly?

Comment (lumirlumir) [reply src/rules/no-reference-like-urls.js:32]: As far as I remember, there were some tricky edge cases that prevented us from directly visiting `link` and `image` nodes.

One edge case was:

https://github.com/eslint/markdown/blob/f80a9e1d6b505e42491f49fc6ac36eeacd3bdbab/tests/rules/no-reference-like-urls.test.js#L82-L91

https://github.com/eslint/markdown/blob/f80a9e1d6b505e42491f49fc6ac36eeacd3bdbab/tests/rules/no-reference-like-urls.test.js#L556-L577

The `link` and `image` nodes can have links wrapped in brackets `<>`, but the AST just represents them as links without the brackets.

It wasn’t easy to handle this case by directly using the `link` and `image` nodes.

---

There’s also an edge case with auto-link literals in CommonMark and GFM mode:

https://github.com/eslint/markdown/blob/f80a9e1d6b505e42491f49fc6ac36eeacd3bdbab/tests/rules/no-reference-like-urls.test.js#L92-L99

https://github.com/eslint/markdown/blob/f80a9e1d6b505e42491f49fc6ac36eeacd3bdbab/tests/rules/no-reference-like-urls.test.js#L100-L107

These cases shouldn’t be reported either, but if we look at the `link` node directly, it can be pretty tricky to handle all of the edge cases correctly.

---

It would be nice to refactor this code to visit `link` and `image` nodes directly, but in my experience, it hasn’t been easy to implement.

So, I've thought the current approach is reasonable, especially since [`no-reversed-media-syntax`](https://github.com/eslint/markdown/blob/main/src/rules/no-reversed-media-syntax.js) and other rules for detecting error-prone syntax use similar strategies.

Comment (mdjermanovic) [reply src/rules/no-reference-like-urls.js:32]: Can we then visit `link` and `image` nodes directly and use the regex on their text to check if they're using the syntax this rule wants to check? I think it would be less error-prone then parsing whole raw paragraphs.

> So, I've thought the current approach is reasonable, especially since [`no-reversed-media-syntax`](https://github.com/eslint/markdown/blob/main/src/rules/no-reversed-media-syntax.js) and other rules for detecting error-prone syntax use similar strategies.

I think `no-reversed-media-syntax` is a different case, because text with reversed syntax is not parsed as a link. Here, in `no-reference-like-urls`, we could use the fact that text we want to check is already parsed as `link`/`image` by the parser, if that's always the case?

Comment (lumirlumir) [reply src/rules/no-reference-like-urls.js:32]: I think your suggestion could help make this rule more reliable 👍

I agree with your idea, but the scope seems a bit large for this particular fix. It will also take me some time to figure out the best approach, so if it’s alright with you, I’d like to create a separate PR to refactor the code.

Comment (mdjermanovic) [reply src/rules/no-reference-like-urls.js:32]: Okay, then let's fix the problem this way. I could try to simplify rule to visit `link` and `image` nodes directly after this is merged.
